vortex-mask/src/lib.rs

```diff
@@ -0,0 +1,416 @@
+        lhs: impl Iterator<Item = usize>,
+        rhs: impl Iterator<Item = usize>,
+    ) -> Self {
+        let mut intersection = Vec::with_capacity(len);
+        let mut lhs = lhs.peekable();
+        let mut rhs = rhs.peekable();
+        while let (Some(&l), Some(&r)) = (lhs.peek(), rhs.peek()) {
+            match l.cmp(&r) {
+                Ordering::Less => {
+                    lhs.next();
+                }
+                Ordering::Greater => {
+                    rhs.next();
+                }
+                Ordering::Equal => {
+                    intersection.push(l);
+                    lhs.next();
+                    rhs.next();
+                }
+            }
+        }
+        Self::from_indices(len, intersection)
+    }
+
+    #[inline]
+    // There is good definition of is_empty, does it mean len == 0 or true_count == 0?
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        self.0.len
+    }
+
+    /// Get the true count of the mask.
+    #[inline]
+    pub fn true_count(&self) -> usize {
+        self.0.true_count
+    }
+
+    /// Get the false count of the mask.
+    #[inline]
+    pub fn false_count(&self) -> usize {
+        self.len() - self.true_count()
+    }
+
+    /// Return the selectivity of the full mask.
+    #[inline]
+    pub fn selectivity(&self) -> f64 {
+        self.0.selectivity
+    }
+
+    /// Get the canonical representation of the mask.
+    pub fn boolean_buffer(&self) -> &BooleanBuffer {
+        self.0.buffer()
+    }
+
+    /// Get the indices of the true values in the mask.
+    pub fn indices(&self) -> &[usize] {
+        self.0.indices()
+    }
+
+    /// Get the slices of the true values in the mask.
+    pub fn slices(&self) -> &[(usize, usize)] {
+        self.0.slices()
+    }
+
+    /// Returns the first true index in the mask.
+    pub fn first(&self) -> Option<usize> {
+        self.0.first()
+    }
+
+    /// Returns the best iterator based on a selectivity threshold.
+    ///
+    /// Currently, this threshold is fixed at 0.8 based on Arrow Rust.
+    pub fn iter(&self) -> Iter {
+        if self.selectivity() > FILTER_SLICES_SELECTIVITY_THRESHOLD {
+            Iter::Slices(self.slices())
+        } else {
+            Iter::Indices(self.indices())
+        }
+    }
+
+    /// Slice the mask.
+    pub fn slice(&self, offset: usize, length: usize) -> Self {
+        if self.true_count() == 0 {
+            return Self::new_false(length);
+        }
+        if self.true_count() == self.len() {
+            return Self::new_true(length);
+        }
+
+        if let Some(buffer) = self.0.buffer.get() {
+            return Self::from_buffer(buffer.slice(offset, length));
+        }
+
+        let end = offset + length;
+
+        if let Some(indices) = self.0.indices.get() {
+            let indices = indices
+                .iter()
+                .copied()
+                .filter(|&idx| offset <= idx && idx < end)
```

Comment:
For tiny slices on big masks with lots of set indices, it seems like we should short-circuit, right?
```suggestion
            .skip_while(|idx| idx < offset)
            .take_while(|idx| idx < end)
```